ql/src/java/org/apache/hadoop/hive/ql/exec/ConditionalTask.java

```diff
@@ -32,6 +33,8 @@
 public class ConditionalTask extends Task<ConditionalWork> implements Serializable {
 
   private static final long serialVersionUID = 1L;
+  private static final ReentrantLock RESOLVE_TASK_LOCK = new ReentrantLock();
```

Comment:
Having a JVM level global lock seems quite problematic. It could stop every query running on this HS2 instance.
Could we find a way to restrict the scope of the lock?

Reply:
That's a good point! I changed it to session scope lock.

Reply:
What about query level?
We can change `queryStateMap` to concurrent, and put the lock into the Query State. What do you think?

Reply:
I think that's nice! And we may just use `Task.queryState` instead of accessing `queryStateMap`?
I put the lock in session level because in our use case we seldom run queries in parallel within one session.